### PR TITLE
SWIFT-904 Add connectTimeoutMS option to MongoClientOptions

### PR DESCRIPTION
**Pull request by bynn**
SWIFT-904 Add connectTimeoutMS option to MongoClientOptions

### CODE REVIEW

File: Sources/MongoSwift/MongoClient.swift
Comment (patrickfreed) [inline Sources/MongoSwift/MongoClient.swift:19]: I like the idea of using a NIO type for this, but I think for consistency with our other timeout options we'll need to continue to use just `Int` here. In 2.0 (the pure swift version), I think we'd probably want to update these various timeouts to use such a type, though I think we might opt for the more general `TimeAmount` instead, which `ChannelOptions.Types.ConnectTimeoutOption` is basically an alias for. Could you file a ticket to remind us to do this in 2.0?

BTW, the reason we have to wait for 2.0 to update the types is that the driver follows [Semantic Versioning](https://semver.org/), and since changing the type of existing options would constitute a breaking change, we'd only be able to introduce it in a new major version. We're currently not planning on doing another major version release until the new driver is done, though.

File: Sources/MongoSwift/MongoClient.swift
Comment (patrickfreed) [inline Sources/MongoSwift/MongoClient.swift:20]: In order for this option to be used by the driver, we'll need to pass it in to libmongoc. The way we do that is via the various `applyAndValidateXOptions` methods in `ConnectionString.swift`. For this option in particular, we'll want to add logic into `applyAndValidateConnectionPoolOptions`.

Once that's done, can we add a test to verify that the setting works? We'll probably want to put it into `SyncMongoClientTests.swift` (so that it will test both sync + async API).

An example test could be something like:
1. enable a [failCommand failpoint](https://github.com/mongodb/mongo/wiki/The-%22failCommand%22-fail-point) on the `isMaster` command with blockTimeMS = 500
    - This will ensure that a connection attempt takes at least 500ms, since the `isMaster` command is always run as part of the handshake for a new connection
2. create a client with `connectTimeoutMS` = 100
3. try to run a `ping` command and assert that a failure occurs
4. create a client with `connectTimeoutMS` = 1000
5. try to run a `ping` command and assert that it succeeds

For an example of a test that makes use of a failpoint, check out [`testCursorTerminatesOnError`](https://github.com/mongodb/mongo-swift-driver/blob/main/Tests/MongoSwiftSyncTests/MongoCursorTests.swift#L243).

In order to set failpoints, be sure to start your `mongod` locally with the `--setParameter enableTestCommands=1` option.

File: Sources/MongoSwift/MongoClient.swift
Comment (patrickfreed) [inline Sources/MongoSwift/MongoClient.swift:18]: Some docs nits:
- can we add "maximum" before "time" here?
- can we change "the" (in "the MongoDB server") to "a"?
- this value actually also applies to monitoring connections (i.e. connections not included in the pools), so can we drop the "from your connection pool"?

**Review by patrickfreed — COMMENTED** [review]
Looks like a good start so far!

A few git related comments:
- you can delete the default checklist in the PR description. We have that here to ensure community members take a few steps before opening PRs, but we don't usually include it ourselves. Instead you can just have the ticket number and a brief description of the changes in the PR, if necessary.
- [Make sure to update the email in your local git client](https://docs.github.com/en/github/setting-up-and-managing-your-github-user-account/setting-your-commit-email-address#setting-your-commit-email-address-in-git) to be the same as the one in your GitHub account so GitHub can associate the commits with you properly.

File: Sources/MongoSwift/MongoClient.swift
Comment (bynn) [reply Sources/MongoSwift/MongoClient.swift:19]: done

File: Sources/MongoSwift/ConnectionString.swift
Comment (kmahar) [inline Sources/MongoSwift/ConnectionString.swift:301]: I think 0 is actually a valid value for this option, according to the [URI options spec](https://github.com/mongodb/specifications/blob/master/source/uri-options/uri-options.rst#list-of-specified-options):
> non-negative integer; 0 means "no timeout"

relatedly, for a lot of these options we have various tests in `ConnectionStringTests.swift` which verify that we e.g. error on invalid values, correctly prioritize options provided via `MongoClientOptions` over those provided via the connection string itself. I think it would be good to add some of those for this option; you might take some inspiration from e.g. `testLocalThresholdMSOption`

File: Tests/MongoSwiftSyncTests/SyncMongoClientTests.swift
Comment (kmahar) [inline Tests/MongoSwiftSyncTests/SyncMongoClientTests.swift:337]: nit: I would put the setupClient initialization and check for failCommand support first in this test before the rest of the code, so we can "fail fast"/exit quickly and the exit conditions are easier to spot at a quick glance.

**Review by kmahar — CHANGES_REQUESTED** [review]
looks great! some minor comments around an edge case and testing.

**Comment by kmahar** [comment]
Oh, also BTW since you added a new test, you need to run `make linuxmain` from the root directory to re-generate the list of tests in `LinuxMain.swift`, which is required for testing on Linux.

File: Tests/MongoSwiftSyncTests/SpecTestRunner/FailPoint.swift
Comment (patrickfreed) [inline Tests/MongoSwiftSyncTests/SpecTestRunner/FailPoint.swift:170]: Given that we'll always want `blockConnection` to be true if we've specified a non-nil `blockTimeMS`, can we drop the `blockConnection` argument to this method and just set it true if `blockTimeMS` is specified?

File: Sources/MongoSwift/MongoClient.swift
Comment (patrickfreed) [inline Sources/MongoSwift/MongoClient.swift:17]:
```suggestion
    /// Specifies the maximum time, in milliseconds, for an individual connection to establish a TCP
```

**Review by patrickfreed — COMMENTED** [review]
Looking good, just a few minor suggestions. BTW, it seems like your email still isn't set. Once it is, your profile picture / GitHub should start showing up next to your commits instead of the default one.

File: Tests/MongoSwiftSyncTests/SyncMongoClientTests.swift
Comment (kmahar) [inline Tests/MongoSwiftSyncTests/SyncMongoClientTests.swift:340]: there looks to be an issue on 4.0 on Evergreen. it looks like blockConnection actually isn't supported on server versions < 4.2, so we should skip this test on those:
https://jira.mongodb.org/browse/SERVER-41070

File: Tests/LinuxMain.swift
Comment (kmahar) [inline Tests/LinuxMain.swift:1]: the sourcery task is going to fail on evg because of this; once my PR #614 is merged in you can rebase on top of it and that should fix it.

File: Sources/MongoSwift/ConnectionString.swift
Comment (kmahar) [inline Sources/MongoSwift/ConnectionString.swift:96]: oh no.... I'm sorry to go back and forth on this, but your change here made me look into why I'd done this in the first place.

 libmongoc does not comply with the spec: it does _not_ treat a value of 0 for any timeout option as "no timeout", but instead treats it as "use the default".

see [here](http://mongoc.org/libmongoc/current/mongoc_uri_t.html#connection-options):
> Setting any of the *TimeoutMS options above to 0 will be interpreted as “use the default value”.

since its behavior when 0 is specified is non-spec compliant, I think it would be better for us to not allow a value of `0` for these options after all, and go back to your original code.

in 2.0 we can change our behavior to be in line with the spec.

I should have opened a ticket about this but I don't think I did. Could you please open a ticket along the lines of "allow users to specify 0 for timeout values"?

File: Tests/MongoSwiftSyncTests/SyncMongoClientTests.swift
Comment (patrickfreed) [reply Tests/MongoSwiftSyncTests/SyncMongoClientTests.swift:340]: Specifically, it looks like 4.2.9 was the first server version to support it.

File: .evergreen/config.yml
Comment (patrickfreed) [inline .evergreen/config.yml:353]: In order to prevent this diff from @kmahar 's PR from showing up here too, can we rebase off of `main` instead of doing a merge commit?

File: Sources/MongoSwift/ConnectionString.swift
Comment (patrickfreed) [reply Sources/MongoSwift/ConnectionString.swift:96]: I think this still needs to be changed back.

File: Sources/TestsCommon/CommonTestUtils.swift
Comment (patrickfreed) [inline Sources/TestsCommon/CommonTestUtils.swift:222]: This option should be fine to work on any topology with that version, so I think we can just omit `acceptableTopologies` here.

File: Tests/MongoSwiftSyncTests/SyncMongoClientTests.swift
Comment (patrickfreed) [inline Tests/MongoSwiftSyncTests/SyncMongoClientTests.swift:330]: Supporting blockTime on fail command implies support for failCommand, so this check can be removed now.

File: Sources/TestsCommon/CommonTestUtils.swift
Comment (bynn) [reply Sources/TestsCommon/CommonTestUtils.swift:222]: done

File: Sources/MongoSwift/ConnectionString.swift
Comment (bynn) [reply Sources/MongoSwift/ConnectionString.swift]: changed back

File: .evergreen/config.yml
Comment (bynn) [reply .evergreen/config.yml]: done

File: Tests/MongoSwiftSyncTests/SyncMongoClientTests.swift
Comment (bynn) [reply Tests/MongoSwiftSyncTests/SyncMongoClientTests.swift]: done

**Review by kmahar — APPROVED** [review]
lgtm assuming Evergreen goes green! nice work, this ended up being a tricky one 🙂

**Review by patrickfreed — APPROVED** [review]
LGTM pending green evg run, great job with this!

**Review by patrickfreed — APPROVED** [review]

**Pull request merged**
